iroh-net/src/magicsock/endpoint.rs

```diff
@@ -203,36 +201,40 @@ impl Endpoint {
-                if !self.is_best_addr_valid(*now) {
-                    // We had a best_addr but it expired so send both to it and DERP.
-                    trace!(addr = %best_addr.addr, latency = ?best_addr.latency, "best_addr is set but outdated, use best_addr and derp");
-                    (Some(best_addr.addr), self.derp_region(), true)
-                } else {
-                    trace!(addr = %best_addr.addr, latency = ?best_addr.latency, "best_addr is set and valid, use best_addr only");
-                    // Address is current and can be used
-                    (Some(best_addr.addr), None, false)
-                }
+        // Update our best addr from candidate addresses (only if it is empty and if we have recent
+        // pongs).
+        self.update_best_addr_from_candidates();
+        match self.best_addr.state(*now) {
+            // we have a valid address: use it!
+            best_addr::State::Valid(best_addr) => {
+                trace!(addr = %best_addr.addr, latency = ?best_addr.latency, "best_addr is set and valid, use best_addr only");
+                (Some(best_addr.addr), None, false)
             }
-            None => {
-                let (addr, should_ping) = self.get_candidate_udp_addr();
-
-                // Provide backup derp region if no known latency or no addr.
-                let derp_region = if should_ping || addr.is_none() {
-                    self.derp_region()
-                } else {
-                    None
-                };
-
-                trace!(udp_addr = ?addr, derp_region = ?derp_region, ?should_ping, "best_addr is unset, use candidate addr and derp");
-                (addr, derp_region, should_ping)
+            // we have an outdated address: use it, but use derp as well.
+            best_addr::State::Outdated(best_addr) => {
+                trace!(addr = %best_addr.addr, latency = ?best_addr.latency, "best_addr is set but outdated, use best_addr and derp");
+                (Some(best_addr.addr), self.derp_region(), true)
+            }
+            // we have no best address: use a random canidate if available, and derp as backup.
+            best_addr::State::Empty => {
+                let addr = self
+                    .direct_addr_state
+                    .keys()
+                    .choose_stable(&mut rand::thread_rng())
+                    .map(|ipp| SocketAddr::from(*ipp));
+                trace!(udp_addr = ?addr, "best_addr is unset, use candidate addr and derp");
+                (addr, self.derp_region(), addr.is_some())
             }
         }
     }
 
-    /// Determines a potential best addr for this endpoint. And if the endpoint needs a ping.
-    fn get_candidate_udp_addr(&mut self) -> (Option<SocketAddr>, bool) {
+    /// Update our best_addr (if empty) with the candidate udp addr with the lowest latency.
+    ///
+    /// Returns true if the address was updated.
+    fn update_best_addr_from_candidates(&mut self) -> bool {
```

Comment:
this return value is never used in the code, we can remove it

Reply:
removed